[Core] Add AuraFlow

sayakpaul opened the pull request:
# What does this PR do?

Adds Aura Flow from Fal. 

Test code:

```python
from diffusers import AuraFlowPipeline
from diffusers.utils import make_image_grid
import torch


pipeline = AuraFlowPipeline(
	"AuraDiffusion/auradiffusion-v0.1a0", 
	torch_dtype=torch.float16
).to("cuda")

images = pipeline(
    prompt="a cute cat with tiger like looks",
    height=512,
    width=512,
    num_inference_steps=50, 
    num_images_per_prompt=4,
    generator=torch.Generator().manual_seed(666),
    guidance_scale=3.5,
).images
make_image_grid(images, 1, 4).save("demo_hf.png")
```

> [!WARNING]  
> To download the model you must be a member of the AuraDiffusion org. Follow this (internal) Slack [message](https://huggingface.slack.com/archives/C07ATBZ47KQ/p1720062535940439?thread_ts=1720062481.505559&cid=C07ATBZ47KQ). 

Gives:

![image](https://github.com/huggingface/diffusers/assets/22957388/0b256155-6d30-4b31-8729-ec69069a7267)

## TODOS

- [ ] Docs
- [x] Tests
- [x] Scheduler (@yiyixuxu would you be able to help out here? I couldn't find a way to use our existing flow matching scheduler in this case)

~Because of the last point above, the noise scheduling code is taken from the original codebase. But I think this PR is still ready for a first review.~

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_8796). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sayakpaul commented on src/diffusers/models/transformers/lavender_transformer_2d.py:
```diff
@@ -0,0 +1,401 @@
+# Copyright 2024 Stability AI, Lavender Flow, The HuggingFace Team. All rights reserved.
```
New model because it differs from the SD3 one:

* Uses register tokens
* Mixes MMDiT and another kind of simple DiT block (that uses a concatenated `encoder_hidden_states` and `hidden_states`) 
* The final layer norm is different 
* Position embeddings are different (uses learned positional embeddings) 
* The feedforward is different. We only support GeLU and its variants in the feedforward. It uses SwiGLU. 
* No pooled projections.

sayakpaul commented on src/diffusers/models/transformers/lavender_transformer_2d.py:
```diff
@@ -0,0 +1,401 @@
+    def _set_gradient_checkpointing(self, module, value=False):
+        if hasattr(module, "gradient_checkpointing"):
+            module.gradient_checkpointing = value
+
```
I have deliberately kept additional methods like feedforward chunking, QKV fusion, etc. out of this class because it helps with the initial review.

sayakpaul commented on src/diffusers/pipelines/lavender_flow/pipeline_lavender_flow.py:
```diff
@@ -0,0 +1,509 @@
+
+        # 6. Denoising loop
+        num_warmup_steps = max(len(timesteps) - num_inference_steps * self.scheduler.order, 0)
+        dt = 1.0 / num_inference_steps
+        dt = (
+            torch.tensor([dt] * effective_batch_size)
+            .to(self.device)
+            .view([effective_batch_size, *([1] * len(latents.shape[1:]))])
+        )
```
@yiyixuxu need your help with the scheduler here.

yiyixuxu replied:
ok will take a look tomorrow

yiyixuxu reviewed (commented):
very nice!
left some comments mainly on attention processor

sayakpaul commented:
Looking into the test failures 👀

sayakpaul commented:
Opened a PR to fix the tests https://github.com/huggingface/diffusers/pull/8800.

s9anus98a commented:
> # What does this PR do?
> 
> Adds Lavender Flow from Fal.
> 
> Test code:
> 
> ```python
> from diffusers import LavenderFlowPipeline
> from diffusers.utils import make_image_grid
> import torch
> 
> 
> pipeline = LavenderFlowPipeline(
> 	"AuraDiffusion/auradiffusion-v0.1a0", 
> 	revision="refs/pr/1",
> 	torch_dtype=torch.float16
> ).to("cuda")
> 
> images = pipeline(
>     prompt="a cute cat with tiger like looks",
>     height=512,
>     width=512,
>     num_inference_steps=50, 
>     num_images_per_prompt=4,
>     generator=torch.Generator().manual_seed(666),
>     guidance_scale=3.5,
> ).images
> make_image_grid(images, 1, 4).save("demo_hf.png")
> ```
> 
> Warning
> 
> To download the model you must be a member of the AuraDiffusion org. Follow this (internal) Slack [message](https://huggingface.slack.com/archives/C07ATBZ47KQ/p1720062535940439?thread_ts=1720062481.505559&cid=C07ATBZ47KQ).
> 
> Gives:
> 
> ![image](https://private-user-images.githubusercontent.com/22957388/346060663-0b256155-6d30-4b31-8729-ec69069a7267.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAyMzY3NTgsIm5iZiI6MTcyMDIzNjQ1OCwicGF0aCI6Ii8yMjk1NzM4OC8zNDYwNjA2NjMtMGIyNTYxNTUtNmQzMC00YjMxLTg3MjktZWM2OTA2OWE3MjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA2VDAzMjczOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExYWFiMzM5Y2Y4NWJhODU4MzExNTM5OGMwNWY0ZTJkNzAyYzYzNTI1NTQ4YjNlM2Y5ODc4ZmIyNWMxYzMyZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Mkqu4FPCk3jLRdQG81DvLKAWZjq6tRgFbLaEkvj6yWs)
> ## TODOS
> 
>     * [ ]  Docs
> 
>     * [ ]  Tests
> 
>     * [ ]  Scheduler (@yiyixuxu would you be able to help out here? I couldn't find a way to use our existing flow matching scheduler in this case)
> 
> 
> Because of the last point above, the noise scheduling code is taken from the original codebase. But I think this PR is still ready for a first review.

i can't join that slack only for huggingface employee? is the model release for public?

sayakpaul commented:
@s9anus98a yeah it's not ready for public testing yet.

DN6 commented on src/diffusers/models/attention_processor.py:
```diff
@@ -117,6 +118,8 @@ def __init__(
         context_pre_only=None,
     ):
         super().__init__()
+        from .normalization import FP32LayerNorm
```
Why import this here?

DN6 commented on src/diffusers/models/transformers/auraflow_transformer_2d.py:
```diff
@@ -0,0 +1,399 @@
+        n_hidden = int(2 * hidden_dim / 3)
+        n_hidden = find_multiple(n_hidden, 256)
+
+        self.c_fc1 = nn.Linear(dim, n_hidden, bias=False)
+        self.c_fc2 = nn.Linear(dim, n_hidden, bias=False)
+        self.c_proj = nn.Linear(n_hidden, dim, bias=False)
```
Could we make the naming here a bit more descriptive? What does "c" mean here?

DN6 commented on src/diffusers/models/transformers/auraflow_transformer_2d.py:
```diff
@@ -0,0 +1,399 @@
+        if hidden_dim is None:
+            hidden_dim = 4 * dim
+
+        n_hidden = int(2 * hidden_dim / 3)
+        n_hidden = find_multiple(n_hidden, 256)
```
Could we make the variable names a bit more descriptive.

DN6 commented on src/diffusers/models/transformers/auraflow_transformer_2d.py:
```diff
@@ -0,0 +1,399 @@
+    def __init__(self, dim, num_attention_heads, attention_head_dim, is_last=False):
+        super().__init__()
+        self.is_last = is_last
+        context_norm_type = "ada_norm_continous" if is_last else "ada_norm_zero"
```
IMO it is better to have `context_norm_type` in the `__init__` We can set it to the appropriate value when creating the blocks

```python
for i in range(self.config.num_mmdit_layers):
     is_last = i == len(self.config.num_mmdit_layers) - 1
     AuraFlowTransformerBlock(context_norm_type="ada_norm_continuous if is_last else "ada_norm_zero")
```

sayakpaul replied:
`is_last` is enough to determine other block initialization, not just this block. It's something we followed for SD3, too. So, I think this should be okay.

DN6 commented on src/diffusers/models/transformers/auraflow_transformer_2d.py:
```diff
@@ -0,0 +1,399 @@
+                AuraFlowTransformerBlock(
+                    dim=self.inner_dim,
+                    num_attention_heads=self.config.num_attention_heads,
+                    attention_head_dim=self.config.attention_head_dim,
```
Here wouldn't it mean the `context_norm_type` would always be `ada_norm_zero`? We're not passing in information about the whether this is the final block? Do we then need the check defined here
https://github.com/huggingface/diffusers/pull/8796/files#r1668210149

sayakpaul replied:
Resolved.

DN6 replied:
Sorry, what I mean here is, why even have `if context_norm_type == "ada_norm_continous"` if it isn't used anywhere?

sayakpaul replied:
Cool. ed3391363 should have addressed it. 

For your next round of review, could I request you to have the comments more on things that might merge blocking (barring documentation)?

DN6 reviewed (commented):
It's looking very nice. I think it's good to proceed with the transformer optimizations, chunked ff etc, and can we add some fast tests?

bghira commented:
this'll need `PeftAdapterMixin` on the transformer module and some analogue to `SD3LoraLoaderMixin`, or, just `SD3LoraLoaderMixin`.

yiyixuxu commented on src/diffusers/models/attention_processor.py:
```diff
@@ -217,6 +223,17 @@ def __init__(
             self.to_add_out = nn.Linear(self.inner_dim, self.out_dim, bias=out_bias)
 
+        if qk_norm is not None and added_kv_proj_dim is not None:
+            if qk_norm == "layer_norm":
+                self.norm_added_q = nn.LayerNorm(dim_head, eps=eps)
+                self.norm_added_k = nn.LayerNorm(dim_head, eps=eps)
```
```suggestion
```
let's add this later when we have a model need this configuration, with joint attention + the regular "layer_norm" qk norm

bghira commented:
i've been testing a fork of this with the LoRA support and it works without any changes to just add the peft adapter to the Transformer2D model and the SD3 LoRA loader mixin to the pipeline.

sayakpaul commented:
@yiyixuxu @DN6 I have addressed the comments. Here are some extended comments from my end: 

* Run conversion again because of changes needed to address https://github.com/huggingface/diffusers/pull/8796#discussion_r1668193938. Here's the Hub PR: https://huggingface.co/AuraDiffusion/auradiffusion-v0.1a0/discussions/3. 
* I think the transformer optimizations can be tackled later in a separate PR. We saw for SD3 that chunked feedforward doesn't lead to a lot of memory saving. So, I would like to first see how much that gain is and then add accordingly. 
* https://github.com/huggingface/diffusers/actions/runs/9850568702/job/27195993322?pr=8796#step:5:173 (circular import for FP32 layer norm). 

@bghira, I will add LoRA support in an immediate future PR once this PR is merged to keep the reviewing scope concrete and manageable.

sayakpaul commented:
I have also added fast tests and decided to make the default value of negative prompt to be None instead of "This is watermark, jpeg image white background, web image". I think this is better aligned with our other pipelines. Will include this negative prompt in the docs once I start adding them.

yiyixuxu commented on src/diffusers/models/normalization.py:
```diff
@@ -205,20 +224,28 @@ def __init__(
+            self.norm = FP32LayerNorm(embedding_dim, eps=eps, elementwise_affine=elementwise_affine, bias=bias)
         elif norm_type == "rms_norm":
             self.norm = RMSNorm(embedding_dim, eps, elementwise_affine)
+        elif norm_type == "no_norm":
```
I don't think we should call it `AdaLayerNormContinuous` without the normalization layer here 
I think we can move this out and make a special layer in `auraflow_transformers_2d`

sayakpaul replied:
Done!

yiyixuxu commented on src/diffusers/models/transformers/auraflow_transformer_2d.py:
```diff
@@ -0,0 +1,388 @@
+
+
+@maybe_allow_in_graph
+class AuraFlowTransformerBlock(nn.Module):
```
maybe `AuraFlowJointTransformerBlock`?

sayakpaul replied:
Done. Changed `AuraFlowDiTTransformerBlock` to `AuraFlowSingleTransformerBlock` as well for consistency.

yiyixuxu commented on src/diffusers/models/transformers/auraflow_transformer_2d.py:
```diff
@@ -0,0 +1,388 @@
+
+
+# Taken from the original aura flow inference code.
+def find_multiple(n: int, k: int) -> int:
+    if n % k == 0:
+        return n
+    return n + k - (n % k)
```
```suggestion
```

yiyixuxu commented on src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py:
```diff
@@ -158,7 +158,12 @@ def scale_noise(
         return sigma * self.config.num_train_timesteps
 
-    def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.device] = None):
+    def set_timesteps(
```
umm I don't think these changed are introduced in this PR

sayakpaul replied:
@yiyixuxu we merged https://github.com/huggingface/diffusers/issues/8799 into the current PR branch. So, the commits will come here. But they still belong to you, so I guess that is okay?

yiyixuxu replied:
you're right, I was confused 😅

yiyixuxu commented on src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py:
```diff
@@ -0,0 +1,505 @@
+                padding="max_length",
+                return_tensors="pt",
+            )
+            text_inputs = {k: v.to(device) for k, v in text_inputs.items()}
```
just curious what else is in `text_inputs` other than the `text_input_ids`?

sayakpaul replied:
`attention_mask`

yiyixuxu commented on src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py:
```diff
@@ -0,0 +1,505 @@
+        negative_prompt_attention_mask: Optional[torch.Tensor] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        callback: Optional[Callable[[int, int, torch.Tensor], None]] = None,
+        callback_steps: int = 1,
```
```suggestion
```
these are deprecated, I think it is ok to add callback support for now, but if we are adding it, we should use the new callback API https://github.com/huggingface/diffusers/blob/a785992c1d6fcb1ff66f8a0d68d09a0a81b909e8/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L773

yiyixuxu reviewed (commented):
very nice!!

yiyixuxu reviewed (approved):
thanks!
